academic-observatory-workflows/academic_observatory_workflows/crossref_metadata_telescope/telescope.py

```diff
@@ -0,0 +1,326 @@
+# Copyright 2020-2024 Curtin University
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Author: Aniek Roelofs, James Diprose, Keegan Smith
+
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+import os
+from typing import Optional
+
+import pendulum
+from airflow import DAG
+from airflow.decorators import dag, task
+from airflow.providers.cncf.kubernetes.secret import Secret
+from kubernetes.client import models as k8s
+from kubernetes.client.models import V1ResourceRequirements
+
+from academic_observatory_workflows.config import project_path
+from observatory_platform.airflow.airflow import on_failure_callback
+from observatory_platform.airflow.tasks import check_dependencies, gke_create_storage, gke_delete_storage
+from observatory_platform.airflow.workflow import CloudWorkspace
+
+
+class _TaskResources:
+    def __init__(self, overrides: Optional[dict] = None):
+        """Describes the resources for task containers
+
+        :param overides: Optionally provide a custom resource definition for a task. For example, to override the
+        defaults for the download task, provide {"download": {"memory": "1G", "cpu": "500m"}}"""
+
+        task_resources = {
+            "download": {"memory": "2G", "cpu": "2000m"},
+            "upload_downloaded": {"memory": "4Gi", "cpu": "4"},
+            "extract": {"memory": "8Gi", "cpu": "8"},
+            "transform": {"memory": "16G", "cpu": "16"},
+            "upload_transformed": {"memory": "8G", "cpu": "8"},
+        }
+        if overrides:
+            for task, resources in overrides.items():
+                task_resources[task] = resources
+
+        self.download = V1ResourceRequirements(requests=task_resources["download"], limits=task_resources["download"])
+        self.upload_downloaded = V1ResourceRequirements(
+            requests=task_resources["upload_downloaded"], limits=task_resources["upload_downloaded"]
+        )
+        self.extract = V1ResourceRequirements(requests=task_resources["extract"], limits=task_resources["extract"])
+        self.transform = V1ResourceRequirements(
+            requests=task_resources["transform"], limits=task_resources["transform"]
+        )
+        self.upload_transformed = V1ResourceRequirements(
+            requests=task_resources["upload_transformed"], limits=task_resources["upload_transformed"]
+        )
```

Comment:
I think it would be better to make something that can be used in a more generic way and for each kubernetes task decorator, e.g.

```python
def make_container_resources(default: dict, override: Optional[dict]) -> V1ResourceRequirements:
    resource = default
    if override is not None:
        resource = override
    return V1ResourceRequirements(requests=resource, limits=resource)
```

```python
@task.kubernetes(
    name="extract",
    container_resources=make_container_resources(
        {"memory": "8Gi", "cpu": "8"}, dag_params.gke_resource_overrides.get("extract")
    ),
    **kubernetes_task_params,
)
def extract(release: dict, **context):
    """Extracts the compressed data"""

    from academic_observatory_workflows.crossref_metadata_telescope.tasks import extract

    extract(release, **context)
```

You just have to change the `gke_resource_overrides` variable to: `gke_resource_overrides: dict = field(default_factory=dict)`